### PR TITLE
[DataGrid] Fix memoized selectors with arguments

### PR DESCRIPTION
**Pull request by MBilalShafi**
As reported by @KenanYusuf, memoized selectors with arguments weren't working properly, this PR aims to fix that.

Extracted from #15200 to separate from the BC.

### CODE REVIEW

File: packages/x-data-grid/src/hooks/utils/useGridSelector.ts [hunk 1]
```diff
@@ -127,13 +128,19 @@ export const useGridSelectorV8 = <Api extends GridApiCommon, Args, T>(
   refs.current.state = state;
   refs.current.equals = equals;
   refs.current.selector = selector;
+  const prevArgs = refs.current.args;
+  refs.current.args = args;
+  if (didInit && !refs.current.equals<typeof args>(prevArgs, args)) {
+    // React to arguments change
+    apiRef.current.store.update(apiRef.current.state);
+  }
```
Comment (romgrk) [inline packages/x-data-grid/src/hooks/utils/useGridSelector.ts:136]: Why is this needed? I find the `store.update` surprising. This should listen to the store, not update it.

Comment (MBilalShafi) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts:136]: It is to recompute the particular substate on the arguments' reference update.

We could skip this, but then the arguments change will not be reflected immediately, only on the next `setState` call which might not be the required behavior in some use cases.
As I understand, `store.update()` will only impact those subscribers whose actual state value gets modified. Ideally, the one using this particular `args` value.

Do you see any obvious drawbacks?

**Review by KenanYusuf — APPROVED** [review]
Tested on my working branch using memoized v8 selectors and it is working as expected 👍

File: packages/x-data-grid/src/hooks/utils/useGridSelector.ts (hunk 1 above)
Comment (romgrk) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts:136]: `store.update()` calls every selector (scales as `O(n)`), I'd rather avoid it if there's no reason to call it.

This is just to update the current selector right? Wouldn't it be more simple to call `setState(applySelector(...etc))`?

And is `refs.current.equals` good for this use-case? Can `args` be an array?

Comment (MBilalShafi) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts:136]: > This is just to update the current selector right?

I am not sure honestly, I can't think of a case where it'd affect any other selector, so yes updating the local state seems to be a better alternative.

> And is `refs.current.equals` good for this use-case? Can args be an array?

99% of the usecases should be simple ones (string, number) or objects, even for arrays the reference comparison should work fine, no?

Comment (MBilalShafi) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts]: I think disallowing object type would reduce use case with more than one argument, and I imagine there will be multiple such use cases (for one of the use-cases, I needed to pass `rowId` and `apiRef`)

I'd rather expect the Object/Array reference to be stabilized by the user (it's mostly the internal team so it should be achievable), we could add a respective docs section when we make this API the default to explain the need to the end users too.

Wdyt?

Comment (romgrk) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts]: It would be unergonomic to stabilize the arg, it would be something like this:

```
useGridSelector(apiRef, selector, useMemo(() => [a, b, c], [a, b, c]))
```

If we're going to support objects/arrays, we should use a custom comparison function that does a shallow comparison for complex types, and `Object.is` for primitive types. Wdyt?

Comment (MBilalShafi) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts]: It's a nice idea, I had the same in the back of my mind when using `refs.current.equals` rather than equality or `Object.is` so that users have the flexibility to pass on a custom `equals` but that would also replace it for state comparison too.

Having a custom internal function for this could be more ergonomic, yes 👍 

What would be the best way to decide which one to use, runtime decision?
```tsx
let argsEqual = Object.is;
if (!['string', 'number', 'boolean'].includes(typeof args)) {
  argsEqual = fastObjectShallowCompare;
}
if (didInit && argsEqual(prevArgs, args)) {
  // recompute state
}
```

This might not work as expected if the args type changes from primitive to complex and vice versa. How about showing a dev warning in that case?

Also, does allowing users to override the equality check make sense too? I'd avoid it at this point.

**Review by KenanYusuf — COMMENTED** [review]
@MBilalShafi Spotted an issue with the types for selectors created using `createSelectorMemoizedV8`:

When calling the selector directly, it seems to require an `args` arg even if additional args are not used by the selector.

Screenshot 2024-10-31 at 14 54 39

Selector:
```ts
export const gridColumnDefinitionsSelectorV8 = createSelectorMemoizedV8(
  gridColumnFieldsSelectorV8,
  gridColumnLookupSelectorV8,
  (allFields, lookup) => allFields.map((field) => lookup[field]),
);
```

(ignore all the selectors being suffixed with v8, it's for testing purposes)

File: packages/x-data-grid/src/hooks/utils/useGridSelector.ts (hunk 1 above)
Comment (romgrk) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts]: Yes but abstract that logic in a function, `function argsEqual(a, b) { ... }`. I would ignore cases where `typeof a !== typeof b`, let the jsdoc explain that constraint. I wouldn't allow overrides.

Comment (MBilalShafi) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts]: > let the jsdoc explain that constraint

Typescript enforces that as long as you provide the types for a specific arguments parameter. So it should be fine.

image

File: packages/x-data-grid/src/hooks/utils/useGridSelector.ts
[the remark this answers is not in the document]
Comment (KenanYusuf) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts:1]: > I guess the tradeoff is liveable since not providing arguments where required should generate runtime errors for the users to fix them anyways.

I agree on this.

It may be a separate issue, but given the following selectors:

```ts
// This selector has a selector arg
export const gridVisibleColumnDefinitionsSelector = createSelectorMemoizedV8(
  gridColumnDefinitionsSelector,
  gridColumnVisibilityModelSelector,
  gridListColumnSelector,
  (columns, columnVisibilityModel, listColumn, listView?: boolean) => {
    if (listView) {
      return listColumn ? [listColumn] : [];
    }
    return columns.filter((column) => columnVisibilityModel[column.field] !== false);
  },
);

// Uses the previous selector, but does not use the selector arg
export const gridVisibleColumnFieldsSelector = createSelectorMemoizedV8(
  gridVisibleColumnDefinitionsSelector,
  (visibleColumns) => visibleColumns.map((column) => column.field),
);
```

I am seeing this error when calling the selector function:

Screenshot 2024-11-04 at 14 00 59

<details>
<summary>Full error</summary>

```
No overload matches this call.
  Overload 1 of 2, '(apiRef: MutableRefObject<{ state: GridStateCommunity; instanceId: { id: number; }; }>, args?: (GridListColDef & { computedWidth: number; })[] | undefined): string[]', gave the following error.
    Argument of type 'boolean | undefined' is not assignable to parameter of type '(GridListColDef & { computedWidth: number; })[] | undefined'.
      Type 'boolean' is not assignable to type '(GridListColDef & { computedWidth: number; })[]'.
  Overload 2 of 2, '(state: GridStateCommunity, instanceId: { id: number; }): string[]', gave the following error.
    Argument of type 'MutableRefObject<GridPrivateApiCommunity>' is not assignable to parameter of type 'GridStateCommunity'.
      Type 'MutableRefObject<GridPrivateApiCommunity>' is missing the following properties from type 'GridStateCommunity': isRtl, dimensions, rows, visibleRowsLookup, and 19 more.ts(2769)
```

</details>

If I change it to the following, the selector function is typed correctly but the param is unused:
```ts
export const gridVisibleColumnFieldsSelector = createSelectorMemoizedV8(
  gridVisibleColumnDefinitionsSelector,
  (visibleColumns, listView?: boolean) => visibleColumns.map((column) => column.field),
);
```

[the remark this answers is not in the document]
Comment (MBilalShafi) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts:1]: I think it's kinda expected. The derived selectors have to have the same arguments (if not more), only then the underlying selectors will be provided with the correct arguments to compute the substate correctly.

`gridVisibleColumnDefinitionsSelector` requires `listView` argument and `gridVisibleColumnFieldsSelector` is derived from it. It should also require `listView` argument for `gridVisibleColumnDefinitionsSelector` provide the correct value for the `visibleColumns` argument.

If multiple selectors require different arguments, we could use an object as argument. For example:

```tsx
// First selector that requires arguments
const baseSelector = createSelectorMemoized(
	selectorX,
	(selectorXOutput, { baseSelectorArgs }) => {
		// compute sub-state and return
		const baseSelectorOutput = computeValue(selectorXOutput, baseSelectorArgs);
		return baseSelectorOutput;
	}
);

// Another selector that derives `baseSelector` and also requires some more arguments
const derivedSelector = createSelectorMemoized(
	baseSelector,
	(baseSelectorOutput, { baseSelectorArgs, derivedSelectorArgs }) => {
		// This function doesn't itself need `baseSelectorArgs` but it should still make them required
		// so that `baseSelectorOutput` is computed properly
		const derivedSelectorOutput = computeValue(baseSelectorOutput, derivedSelectorArgs);
		return derivedSelectorOutput;
	}
);

// Another selector that derives `derivedSelector` and doesn't require any arguments
const derivedSelector2 = createSelectorMemoized(
	derivedSelector,
	(derivedSelectorOutput, { baseSelectorArgs, derivedSelectorArgs }) => {
		// This function doesn't require any args but it should still `baseSelectorArgs` and `derivedSelectorArgs` make them required
		// so that the underlying selectors are computed properly
		const derivedSelector2Output = computeValue(derivedSelectorOutput);
		return derivedSelector2Output;
	}
);
```

[the remark this answers is not in the document]
Comment (KenanYusuf) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts:1]: I guess it makes sense, but in scenarios where the selector arg is not used, do we just disable `@typescript-eslint/no-unused-vars` on the line of the unused param?

[the remark this answers is not in the document]
Comment (KenanYusuf) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts:1]: It feels quite verbose doing this for each selector that is derived from another with args. 

Ideally, `derivedSelector2` in the example above would include the typed selector args in the function signature based on the selectors that are passed to it.

The same way that TS knows here that this first param is `GridListColDef[]` because it has `gridVisibleColumnDefinitionsSelector` as the first argument of `createSelectorMemoizedV8`:
 
Screenshot 2024-11-05 at 09 36 21

I haven't fully wrapped my head around the types for these selector creator functions, so I don't know if it is possible, but that would be the ideal DX I think.

[the remark this answers is not in the document]
Comment (MBilalShafi) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts:1]: You are right, it's a bit verbose. It aligns with the choice we made in the selectors with arguments PR, i.e. we should explicitly type the arguments when defining the function, but in this case 

Having the underlying selectors' output function's arguments type detected automatically won't be possible without the [two-step generics](https://github.com/mui/mui-x/pull/14236#discussion_r1719929660) as it's only possible to infer the return type of the function.

I think we could introduce an ignore pattern for the `@typescript-eslint/no-unused-vars` rule:

```diff
diff --git a/.eslintrc.js b/.eslintrc.js
index d7b85f4b16..6790094660 100644
--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -148,6 +148,12 @@ module.exports = {
         ),
       },
     ],
+    "@typescript-eslint/no-unused-vars": [
+      "warn",
+      {
+        "argsIgnorePattern": "^_",
+      }
+    ],
     // TODO move rule into the main repo once it ha
s upgraded
     '@typescript-eslint/return-await': 'off',
     'no-restricted-imports': 'off',
```

To allow something like following without the need to add the ignore directive.
(Typescript compiler does it automatically.)

```tsx
const derivedSelector2 = createSelectorMemoized(
	derivedSelector,
	(derivedSelectorOutput, _args: Args) => {
		return `${derivedSelectorOutput}`
	}
);
```

How does that sound?

[the remark this answers is not in the document]
Comment (KenanYusuf) [reply packages/x-data-grid/src/hooks/utils/useGridSelector.ts:1]: Thanks for the context from the initial discussion, makes sense. I think the underscore pattern is a good compromise given the complexities of the alternate solutions. I'm going to continue testing these changes today 👍

**Review by KenanYusuf — APPROVED** [review]
Not sure what's going on with Argos 🤔 but the changes LGTM.

**Comment by MBilalShafi** [comment]
> Not sure what's going on with Argos 🤔 but the changes LGTM.

Yep this is a flaky one, I didn't dig deep yet though it seems to be related to #14130

File: .eslintrc.js [hunk 2]
```diff
@@ -148,6 +148,10 @@ module.exports = {
         ),
       },
     ],
+    '@typescript-eslint/no-unused-vars': [
+      'error',
+      { vars: 'all', args: 'after-used', ignoreRestSiblings: true, argsIgnorePattern: '^_' },
+    ],
```
Comment (MBilalShafi) [inline .eslintrc.js:154]: This change extends the eslint-config-airbnb-typescript [rule](https://github.com/airbnb/javascript/blob/6499695ac11c4640ed0f77f8865a1adcb32d3239/packages/eslint-config-airbnb-base/rules/variables.js#L54) previously in action to add an `argsIgnorePattern` for allowing `_` based unused variables ignored by eslint parser.

CC @mui/x

Comment (michaldudak) [reply .eslintrc.js:154]: :+1: feel free to open a PR to add this to the Core config.
